rfc/system/5606-json-idl.md

```diff
@@ -0,0 +1,343 @@
+# JSON IDL
+
+**Authors:**
+
+- [@Han-Ru](https://github.com/future-outlier)
+- [@Ping-Su](https://github.com/pingsutw)
+
+## 1 Executive Summary
+
+Use json byte string in protobuf instead of json string to fix int is not supported in protobuf struct.
+
+
+## 2 Motivation
+
+In Flytekit, when handling dataclasses, Pydantic base models, and dictionaries, we store data using JSON strings within struct protobuf. This approach causes issues with integers, as protobuf does not support int types, leading to their conversion to floats. This results in performance issues since we need to recursively iterate through all attributes/keys in dataclasses and dictionaries to ensure floats types are converted to int.
+
+Note: We have more than 10 issues about dict, dataclass and Pydantic.
+
+This feature can solve them all.
+
+## 3 Proposed Implementation
+### Flytekit Example
+#### Before
+```python
+@task
+def t1() -> dict:
+  ...
+  return {"a": 1} -> protobuf Struct {"a": 1.0}
+
+@task
+def t2(a: <dict>):
+  print(a["integer"]) # wrong, will be float point
+```
+#### After
+```python
+Json = "json"
+
+@task
+def t1() -> Annotated[dict, Json]: // Json Byte Strings
+  ...
+  return {"a": 1} -> protobuf Json b'{"a": 1}'
+
+@task
+def t2(a: <dict>):
+  print(a["integer"]) # wrong
+```
+
+#### Note
+- We use Annotated[dict, Json]  instead of dict to ensure backward compatibility.
+    - This helps us avoid breaking changes.
```

Comment:
As a user, I wouldn't want to use `Annotated[dict, Json]` to be honest :/

Reply:
Do you think it would be possible that we "put some kind of metadata into the Literal" that "tells the type engine to use the new mechanism"?

Reply:
@Future-Outlier is there any way to just make the reading backwards compatible?  if this happens to read a cached value from an upstream execution that written with the old code, fall back to reading the old code (and just log a warning).

I think most people update admin & propeller at the same time.  If they just upgrade flytekit with this new type, it'll fail at registration time right?  if they just upgrade flytekit & admin, but not propeller, it'll fail at launch time?

Reply:
@fg91 @wild-endeavor 
For flytekit, there's defintely a way to make it backward compatible by adding metadata.

Reply:
> it'll fail at registration time right?
If we specify with default input, and propeller didn't upgrade (take sandbox as an example)
propeller will crash. (This is definitely a bug in propeller)
```python
from flytekit import task, workflow, ImageSpec

flytekit_hash = "4795f8ef3291e5f90730c0a030025ec6bc189190"
flyteidl_hash = "c1182565d62b4885156454da9d0a100b2facf313"

json_idl = f"git+https://github.com/flyteorg/flyte.git@{flyteidl_hash}#subdirectory=flyteidl"
flytekit = f"git+https://github.com/flyteorg/flytekit.git@{flytekit_hash}"

image = ImageSpec(
    packages=[flytekit, json_idl],
    apt_packages=["git"],
    registry="localhost:30000",
)

@task(container_image=image)
def t2(input: dict) -> dict:
    print(type(input["a"]))
    print(type(input["a"]))
    return input

@workflow
def wf(input: dict={'a': 1, }) -> dict:
    return t2(input=input)
```

<img width="1082" alt="image" src="https://github.com/user-attachments/assets/a908648d-0748-4c1a-b1e1-f2cc6137a3b2">


<img width="1260" alt="image" src="https://github.com/user-attachments/assets/247fcc1e-bcd3-46be-a64a-1d33a95bb791">

Reply:
> if they just upgrade flytekit & admin, but not propeller, it'll fail at launch time?
It'll fail before launch time, but I also think it'll fail at launch time too.
The worst part is that the propeller will crash.
Do you think it's ok?
We didn't have good error handling about handling unknown literal types registered to the propeller's compiler should return a good error message response to us, instead out propeller will just crash.